# 🔨 Prepare Libraries


In [ ]:
from tensorflow import keras

import joblib
import numpy as np


# 🏗️ Load the Model & Vectorizer


In [ ]:
model = keras.models.load_model("history/best_model_C.keras")

with open("history/vectorizer_C.pkl", "rb") as file:
    vectorizer = joblib.load(file)


# 📱 Prepare the Test Reviews


In [ ]:
reviews = [
    # 💚 Positive:
    "ELSA Speak rlly helped me fix my accent 😭✨ i sound soooo much better now!!!",
    "Ok this app is legit 🔥🔥 I actually enjoy pronouncing words now lol",
    "Very smooth UI + clear feedback 👍 totally worth the price tbh",
    "Been using for 2 weeks & my coworkers say I sound 'less shy',, love it",
    "ELSA literally became my daily routine 😍 the AI feels like a real coach.",
    "beautiful interface and i like how it explains each sound in detail.",
    "finally an app that doesn't just -listen- but actually TEACHES 😭👏",
    "Used it for job prep--guess who nailed the interview 😏💼 thank u ELSA",
    "works fine so far. my pronunciation sounds less robotic now wkwk",
    "love the way it tracks tiny mistakes makes me wanna keep trying!",
    # 💢 Negative:
    "ughhh keeps crashing on my old phone 😤 pls fix devs",
    "mic never picks up my voice right...kinda useless rn",
    "not worth it. they push premium every 3 mins 😑",
    "feels repetitive after some days, got bored ngl",
    "the scores are random??? say the same word twice, diff score",
    "honestly the free version does almost nothing.",
    "update broke everything. can't even login anymore.",
    "voice feedback sounds robotic and wrong half the time 🫠",
    "money money money ads",
    # 🌀 Ambiguous:
    "one day it's great, next day it's glitchy af.. consistency please 😩",
    "hmm idk if im improving or just getting used to its voice 😅",
    "Let's try. Hope they are not ruining it with annoying ads like the ones next door.",
    "I like it... but also kinda hate it?? Like it motivates me but drains me too 😭",
    "honestly feels smart but dumb at the same time",
    "Looks premium, but acts budget.",
]


# ⚙️ Transform Texts Using TF-IDF


In [ ]:
X_test = vectorizer.transform(reviews)


# 🪄 Predict Sentiment


In [ ]:
predictions = model.predict(X_test)

predicted_labels = np.argmax(predictions, axis=1)
confidence = np.max(predictions, axis=1)

sentiment_map = {0: "Negative 💢", 1: "Positive 💚"}


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 97ms/step


# 📊 Display Results


In [ ]:
print("=== ELSA Speak Review Sentiment Analysis ===\n")

for i, (text, label, conf) in enumerate(zip(reviews, predicted_labels, confidence)):
    print(f"{i+1:02d}. {sentiment_map[label]} — Confidence: {conf:.3f}\n    {text}\n")


=== ELSA Speak Review Sentiment Analysis ===

01. Positive 💚 — Confidence: 0.791
    ELSA Speak rlly helped me fix my accent 😭✨ i sound soooo much better now!!!

02. Positive 💚 — Confidence: 0.840
    Ok this app is legit 🔥🔥 I actually enjoy pronouncing words now lol

03. Positive 💚 — Confidence: 0.789
    Very smooth UI + clear feedback 👍 totally worth the price tbh

04. Positive 💚 — Confidence: 0.939
    Been using for 2 weeks & my coworkers say I sound 'less shy',, love it

05. Positive 💚 — Confidence: 0.702
    ELSA literally became my daily routine 😍 the AI feels like a real coach.

06. Positive 💚 — Confidence: 0.852
    beautiful interface and i like how it explains each sound in detail.

07. Positive 💚 — Confidence: 0.622
    finally an app that doesn't just -listen- but actually TEACHES 😭👏

08. Positive 💚 — Confidence: 0.957
    Used it for job prep--guess who nailed the interview 😏💼 thank u ELSA

09. Positive 💚 — Confidence: 0.888
    works fine so far. my pronunciation sounds

### Results

* **Correct positive:** 10/10
* **Correct negative:** 9/9

### Ambiguous Results

| #  | Correct  | Prediction   |
| -- | -------- | ------------ |
| 20 | Negative | 92% Positive |
| 21 | Neutral  | 89% Positive |
| 22 | Neutral  | 50% Positive |
| 23 | Mixed    | 73% Positive |
| 24 | Mixed    | 89% Negative |
| 25 | Negative | 63% Positive |
